application/forms/Config/UserBackend/LdapBackendForm.php

```diff
@@ -265,38 +242,92 @@ public function createElements(array $formData)
                 'formnovalidate'    => 'formnovalidate'
             )
         );
+    }
+
+    public function isValidPartial(array $formData)
+    {
+        $isAd = isset($formData['type']) && $formData['type'] === 'msldap';
+        $baseDn = null;
+        $hasAdOid = false;
+        $discoverySuccessful = false;
 
-        if ($this->getElement('btn_discover_domain')->isChecked() && isset($formData['resource'])) {
-            $this->populateDomain(ResourceFactory::create($formData['resource']));
+        if (! $isAd && ! empty($this->resources) && isset($formData['discovery_btn'])
+            && $formData['discovery_btn'] === 'discovery_btn') {
+            $discoverySuccessful = true;
+            try {
+                $capabilities = $this->getLdapCapabilities($formData);
+                $baseDn = $capabilities->getDefaultNamingContext();
+                $hasAdOid = $capabilities->isActiveDirectory();
+            } catch (Exception $e) {
+                $this->warning(sprintf(
+                    $this->translate('Failed to discover the chosen LDAP connection: %s'),
+                    $e->getMessage()
+                ));
+                $discoverySuccessful = false;
+            }
         }
+
+        if ($discoverySuccessful) {
+            if ($isAd || $hasAdOid) {
+                // ActiveDirectory defaults
+                $userClass = 'user';
+                $filter = '!(objectClass=computer)';
+                $userNameAttribute = 'sAMAccountName';
+            } else {
+                // OpenLDAP defaults
+                $userClass = 'inetOrgPerson';
+                $filter = null;
+                $userNameAttribute = 'uid';
+            }
+
+            $formData['user_class'] = $userClass;
```

Comment:
@Al2Klimov Please change $formData['filter'] and $formData['base_dn'] only if the user did not enter anything yet. The filter is currently only a hardcoded suggestion for AD and therefore only applicable if the user did not choose one manually yet. The base_dn manually entered by the user is probably more correct especially due to the fact that there is actually something entered.